### DISCUSSION TITLE
Better configuration/settings for setting git user settings for Work or Personal

### DISCUSSION THREAD

**Discussion by PureKrome** [git-for-windows/git#4058] [General]
👋🏻 G'Day all,

## Problem
I have different username/email settings for **WORK** or **PERSONAL** projects.

I usually put all personal git repo's under a single folder called `Personal`.
The same goes for work projects, under a `Work` folder.

I'm finding the current way to configure your git user settings isn't easy and more importantly, doesn't seem to be 100% accurate all the time (aka. buggy?)

I'm _hoping_ this could be a discussion to:
- see if people agree with my assertion
- see if a new proposal can be achieved? or if not (x-plat considerations, etc) at least maybe make it less buggy? Or even, prove that it's a bit buggy?

Currently, I've [documented how I do this on Windows](https://gist.github.com/PureKrome/8bba4fd9cf586eb32f720a957a447ed2).

TL;DR: is that the _global_ user settings need to include an ugly (well, ugly to me) `includeIf` hack. And what makes it ever worse, is that the _syntax_ for the hack I feel is also a bit tough (yeah, I agree that's just an opinion and not factual) and DEFINATELY easy to be error prone, like my first initial attempts...

Here's a quick summary of what I need to do:

`.gitconfig` in my `Home` directory (so on _my_ Windows11 machine: `C:\Users\purek`)
```
[user]
	email = global.common@email.com
	name = Global common name

[includeIf "gitdir/i:c:/projects/work/"]
path = c:/projects/work/.gitconfig

[includeIf "gitdir/i:c:/projects/personal/"]
path = c:/projects/personal/.gitconfig
```

☝🏻 Urgh at the lack of spaces required in the syntax -and- the linux specific path syntax, even when on windows.

`C:\Projects\Work\.gitconfig`
```
[user]
    name = Work Name
    email = work.email@something.com
```

`C:\Projects\Personal\.gitconfig`
```
[user]
    name = Personal alias/Name
    email = personal.email@something.com
```

(or we could omit the custom personal section, which just falls back to the default settings in the home directory).

### Suggestion

Ditch the `includeIf` in the "home" `.gitconfig`. If there's a `.gitConfig` in any folder, any repo's under that "inherit" those settings.

This is _exactly_ what NuGet does with their `nuget.config` settings. If you put a `nuget.config` file in `C:\Projects\Work` then any _child/sub_ folders here, will _also_ inherit those settings. These same _child/sub_ folder do NOT need their own `nuget.config` file -> it's 🪄 magically inherited! But, if any of these _child/sub_ folders do have their _own_ `nuget.config` files, then they will take preferences over what was inherited and _merge_ any settings, with the lowest hierarchy taking preference.

Office NuGet doc reference:
- https://learn.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior?source=recommendations#how-settings-are-applied

This is really simple and easy. Win .. Win!

## Current way is Buggy?

I think the current logic is causing some unintentional bug? When I look at my git history for whatever repo in my `C:\Projects\Work` folder, I sometimes see the _correct_ settings (i.e. my Work name and email). But then i also see a scattering of other commits which have the correct email but the _incorrect_ name! It's like it's my _default_ system wide _name_. I also very rarely do commits via the GitHub UI .. so I'm very confident it's not via that. Also, this is all on one machine and no one else is committing with my name/credentials/etc. Hell no!

---

Is there anyone else using git, like this? Are you finding this is not working (that easily/nicely) for you, too?

Final note: I'm not sure where the best place to have this discussion, so I hope this is ok/the right spot.

**Comment by PhilipOakley** [comment]
A drive by comment. 

Could you say which terminal/shell you are using? It maybe that the quoting requirements vary between shells/terminals, and those quoting methods you are expecting. The quoting problem keeps coming up in a number of threads, so I thought it worth getting those details, along with asking that you also complete as much of the issue template as possible as they do provide extra insight into your setup.

If you can do an [MVCE](https://stackoverflow.com/help/minimal-reproducible-example) as well, then that would be useful.

**Comment by PureKrome** [comment]
> Could you say which terminal/shell you are using?

Um, I'm not to sure. Because, I feel like this occurs in multiple places/programs. I did mentioned in the OP that this is on Windows11.

I use the normal `Windows Terminal` to run git commands but I also use the `Developer Powershell` terminal inside of Visual Studio:

![image](https://user-images.githubusercontent.com/899878/194784251-15f543e3-4e83-4fc4-a9af-fd555ad2d750.png)

Finally, I also run git commands (like commit and push) from inside Visual Studio .. so i'm not sure if that uses a shell also? 

> complete as much of the issue template as possible

Ah, I didn't create an issue because I wasn't sure if this problem I'm having is legit and other people agree. So I haven't looked at what the issue template is. I'll check it out now:

----

### Setup

> Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

64bit

```
$ git --version --build-options

git version 2.33.0.windows.2
cpu: x86_64
built from commit: 8735530946cced809cc6cc4c2ca3b078cdb3dfc8
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

> Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

Edition	Windows 11 Pro
Version	22H2
Installed on	‎1/‎10/‎2022
OS build	22621.608
Experience	Windows Feature Experience Pack 1000.22634.1000.0

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.22621.608]
```

> What options did you set as part of the installation? Or did you choose the defaults?

I can't remember. Usually defaults.

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"

Editor Option: VIM
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled

```
### Details

> Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other
 
Not too sure. I run commands in the `Windows Terminal` (which defaults to Powershell), in the Visual Studio `Developer Powershell` and using the VS Git client.

> What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

I'm not sure -exactly- which command is causing this. I'm _guessing_ it's the `commit` command because that would record my name and email.


> What did you expect to occur after running these commands?

The "inherited" `[user]` section data. Not the default custom-settings. At least, not the default `name`.

> What actually happened instead?

History is showing a mix of `name` values for the commit author. some are from this inherited parent folder (which is good), other's showing my default.

> If the problem was occurring with a specific repository, can you provide the URL to that repository to help us with testing?

Nope. Any of my private work ones.

---


Final thoughts: I also do `SQUASH AND MERGE` in GitHub ... so I'm not sure if this might be doing it? Not sure how to confirm if this is a `git` issue. That said, the main idea of this discussion isn't necessarily about a potential bug, but more about how to make setting custom `[user]` sections way more nicely ... like how the NuGet team do for their custom settings.
